### PR TITLE
[jira] IDEA-149504 add support for Jira Server Personal Access Tokens

### PR DESCRIPTION
**Pull request by thetoothpick**
Adds support for Personal Access Tokens on Jira Server / Datacenter (not Jira Cloud).  This uses the `Authorization: Bearer` HTTP header.

Personal Access Tokens for Jira are documented here: https://confluence.atlassian.com/enterprise/using-personal-access-tokens-1026032365.html#UsingPersonalAccessTokens-UsingPATs

### Implementation notes:

Added support for bearer tokens to the `BaseRepository` and subclasses, including a new repository field, form checkbox, and handlers.  Checking the token checkbox will hide the username field and change the password label to "API Token".

The HTTP library used by the task services doesn't natively support bearer tokens authentication, so this PR disables built-in authentication and explicitly adds the header to requests.

Tested manually against an internal Jira server.

### Screenshot

jira-server-api-token

### CODE REVIEW

**Review by ms-zlab — APPROVED** [review]

**Comment by thetoothpick** [comment]
@ms-zlab I created an alternative implementation which can be packaged as a standalone plugin and installed in IntelliJ (version 231.8109.175, the latest official release): https://github.com/thetoothpick/intellij-community/tree/231.8109-support-jira-server-token-alt.  It works slightly differently than above, you need to uncheck the new "Use HTTP Authentication" box and the password field will change to "API Token".

I uploaded the jar file [here](https://github.com/thetoothpick/intellij-community/releases/tag/231.8109.175-tasks-plugin), download it, install it from disk, and restart IntelliJ to use.

**Comment by axute** [comment]
Fantastic, works perfectly!
Thank you for the preview ;-)

**Comment by ms-zlab** [comment]
Is there anything new? I'm still interested, it works.

**Comment by axute** [comment]
> @ms-zlab I created an alternative implementation which can be packaged as a standalone plugin and installed in IntelliJ (version 231.8109.175, the latest official release): https://github.com/thetoothpick/intellij-community/tree/231.8109-support-jira-server-token-alt. It works slightly differently than above, you need to uncheck the new "Use HTTP Authentication" box and the password field will change to "API Token".
> 
> I uploaded the jar file [here](https://github.com/thetoothpick/intellij-community/releases/tag/231.8109.175-tasks-plugin), download it, install it from disk, and restart IntelliJ to use.

Could you trigger a current release? because it stops working with the current version  IU-231.8770.65

**Comment by dawei-nava** [comment]
Hmm, any progress on merge this PR?

**Comment by thetoothpick** [comment]
I just rebased on latest master.

@east825 are you able to review this?  I see you've been assigned to the PR.

**Comment by thetoothpick** [comment]
@axute I also rebased and rebuilt the plugin JAR for the latest release: https://github.com/thetoothpick/intellij-community/releases/tag/231.9011.34-tasks-plugin

**Comment by Ristee** [comment]
Hello, I can't install plugin now.

> Plugin 'Task Management' (version '231.9011.34') is not compatible with the current version of the IDE, because it requires build 231.9011.34 or older but the current build is PS-231.9011.38

But on previous version (RC) 231.9011.34 it writes opposite
> Plugin 'Task Management' (version '231.9011.34') is not compatible with the current version of the IDE, because it requires build 231.9011.34 or newer but the current build is PS-231.9011.20

Could you help? Thank you.

**Comment by east825** [comment]
Hi! Sorry for such a delay, I have had a couple of business trips. I will review the PR once I finish with ongoing internal reviews before the upcoming feature freeze (hopefully until the end of the next week).

**Comment by pacecreative** [comment]
Is there a chance to get a build that would work with the latest Pycharm update for testing?

**Comment by east825** [comment]
I have to admit that back when I was actively involved in Task Management subsystem development and the issue first appeared, I had been overthinking the problem. I imagined a tighter integration with JIRA's OAuth flow, similar to what we do now for Space, when an IDE opens the browser with the confirmation page and then automatically accepts the authorization token without the need to manually copy and paste anything. Given that the API for directly managing PATs didn't exist then and people also mentioned all kinds of authentication providers in the ticket, it wasn't obvious that it could be easily implemented, and the issue has been getting postponed.

Apparently, though, now with the new PAT controls in JIRA, this extra setting is actually just fine for most use cases. Hats off to @thetoothpick for coming forward and implementing the thing! 👏

**Comment by east825** [comment]
I have a couple of minor comments about the UI and confining the changes to JIRA connector implementation, but apart from those, it's good to go upstream.

**Comment by thetoothpick** [comment]
> Apparently, though, now with the new PAT controls in JIRA, this extra setting is actually just fine for most use cases. Hats off to @thetoothpick for coming forward and implementing the thing! 👏

Thanks!  My first pass at this was writing an `nginx` proxy to add the header, so I figured it couldn't be to hard to do it in code, and decided to give it a shot 🙂.


> I have a couple of minor comments about the UI and confining the changes to JIRA connector implementation, but apart from those, it's good to go upstream.

@east825 I don't see any review comments yet, but regarding confining the changes to the JIRA connector, I have an alternative implementation here that might be closer aligned: https://github.com/thetoothpick/intellij-community/commit/e98b096636a6f9f655068d32b3a97b94def2f8f6.  This was written to allow patching the plugin without modifying core JARs.  The UI is a little more awkward IMO, but I can try to merge the two approaches and make the UI clearer, if that's preferred.

**Comment by thetoothpick** [comment]
> Is there a chance to get a build that would work with the latest Pycharm update for testing?

Unfortunately it's kind of a pain for me to build these on my laptop, but it sounds like this might get merged sooner than later 😄.

**Comment by pacecreative** [comment]
> > Is there a chance to get a build that would work with the latest Pycharm update for testing?
> 
> Unfortunately it's kind of a pain for me to build these on my laptop, but it sounds like this might get merged sooner than later 😄.

No worries! Appreciate all the work to do it, regardless! If I was better in Java I would offer to help. 🤣

**Review by east825 — CHANGES_REQUESTED** [review]

**Comment by east825** [comment]
> @east825 I don't see any review comments yet, but regarding confining the changes to the JIRA connector, I have an alternative implementation here that might be closer aligned: [thetoothpick@e98b096](https://github.com/thetoothpick/intellij-community/commit/e98b096636a6f9f655068d32b3a97b94def2f8f6). This was written to allow patching the plugin without modifying core JARs. The UI is a little more awkward IMO, but I can try to merge the two approaches and make the UI clearer, if that's preferred.

Oh, sorry, I didn't release that the review was still in the pending state. 

I agree that even though piggybacking on the existing "Use HTTP authentication" checkbox allows keeping the changes to the minimum, people might have a hard time figuring out that they have to uncheck it to allow PATs for authentication. Let's have a dedicated checkbox as you did before.

File: platform/tasks-platform-api/resources/messages/TaskApiBundle.properties
Comment (east825) [inline platform/tasks-platform-api/resources/messages/TaskApiBundle.properties:14]: Another bikeshedding thing is that according to our UI guidelines, checkbox labels should use sentence-style capitalization, i.e. "Use personal access token". "Login Anonymously" actually violates it, I will update it.

**Comment by thetoothpick** [comment]
@east825 thanks for the review/feedback, I've pushed a commit to move everything back into the `intellij.tasks.jira` package (except the `TaskApiBundle.properties` file, I wasn't able to find an equivalent in the Jira plugin).

I added a new commit for easier review, it mostly undoes my first commit.  Happy to squash them if desired before merging.

New screenshot:

Screenshot 2023-06-29 at 3 09 24 PM

Cloud URL (showing that that the new checkbox is not displayed in that case):

Screenshot 2023-06-29 at 3 09 49 PM

File: plugins/tasks/tasks-core/jira/src/com/intellij/tasks/jira/JiraRepository.java
Comment (east825) [inline plugins/tasks/tasks-core/jira/src/com/intellij/tasks/jira/JiraRepository.java:281]: I think we still need to have `setUseHttpAuthentication(true)` in an `else` branch here to retain the old behavior when the checkbox is off.

**Review by east825 — CHANGES_REQUESTED** [review]

File: plugins/tasks/tasks-core/jira/src/com/intellij/tasks/jira/JiraRepository.java
Comment (thetoothpick) [reply plugins/tasks/tasks-core/jira/src/com/intellij/tasks/jira/JiraRepository.java:281]: good call, added

**Comment by east825** [comment]
@thetoothpick LGTM! If you want to squash everything before merging, feel free to do so, and I will integrate the changes.

**Comment by thetoothpick** [comment]
> @thetoothpick LGTM! If you want to squash everything before merging, feel free to do so, and I will integrate the changes.

Thanks, squashed.  Is it possible to get this into 2023.2 as well?  (I think that's been branched already?)

**Comment by east825** [comment]
@thetoothpick Right, unfortunately, we're already past the feature freeze and it's a UI-affecting change that needs to be documented and pass the QA scrutiny 😓

**Review by east825 — APPROVED** [review]

**Comment by east825** [comment]
@thetoothpick Merged! A bot should mark the PR as closed soon. Thanks again for contributing! (and patience with the review)

**Comment by east825** [comment]
Closing manually as "closes https://github.com/JetBrains/intellij-community/pull/2401" wasn't added automatically by the script. The corresponding [commit](https://github.com/JetBrains/intellij-community/commit/3457be24bc3a39d632073f80bb3b75d8c7b9c8b0) in the upstream.

**Pull request closed without merging**
